add Debian packaging for python-twodict

nodiscc opened the pull request:
Related to https://github.com/MrS0m30n3/youtube-dl-gui/pull/189

The best and recommended practice is to only push the `debian/` directory to a separate branch (eg. `debian-packaging`), so that it does not interfere with your master branch. When you want to release a new package, just `git checkout debian-packaging; git merge master`, bump the package version/changelog/etc and rebuild.

But Github doesn't let me create a Pull Request for a _new branch_, so you'd have to merge this manually:

```
git remote add nodiscc https://github.com/nodiscc/twodict
git remote update
git checkout debian-packaging
git push -u origin debian-packaging
```

If you want to maintain the .deb packages in the future (I don't know if/how much you use Debian) You may also want to edit the `Maintainer:` field in debian/control. You should also create a GnuPG key for package signing. Otherwise let me know and I'll try to keep them updated.

nodiscc committed:
add Debian packaging for python-twodict
add short build documentation

nodiscc commented:
@MrS0m30n3 ok to merge?

nodiscc commented:
@MrS0m30n3 I can also move _everything_ (including twodict packaging) to a `packaging/debian/` directory in https://github.com/MrS0m30n3/youtube-dl-gui/ if you don't want debian directories in all your projects. Just tell me what you prefer and I'll adapt the patch.

MrS0m30n3 commented on debian/control:
```diff
@@ -0,0 +1,17 @@
+Section: python
+Priority: optional
+Maintainer: nodiscc <nodiscc@gmail.com>
+Build-Depends: debhelper (>= 9), dh-python, python-all
```
You should also add `python3-all` since twodict supports Python 3.*

MrS0m30n3 commented on debian/control:
```diff
@@ -0,0 +1,17 @@
+Priority: optional
+Maintainer: nodiscc <nodiscc@gmail.com>
+Build-Depends: debhelper (>= 9), dh-python, python-all
+X-Python-Version: >= 2.7
```
You should also add `X-Python3-Version: >= 3.3`

MrS0m30n3 commented on debian/control:
```diff
@@ -0,0 +1,17 @@
+
+Package: python-twodict
+Architecture: all
+Depends: ${shlibs:Depends}, ${misc:Depends}, ${python:Depends}
```
\+ `${python3:Depends}`

MrS0m30n3 commented on debian/rules:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/make -f
+
+%:
+	dh $@ --with python2
```
Use **pybuild**, check [here](https://wiki.debian.org/Python/LibraryStyleGuide?action=show&redirect=Python#debian.2Frules)

MrS0m30n3 replied:
Also, make sure the tests pass both in Python2 & 3

MrS0m30n3 reviewed (commented):
Excellent job @nodiscc, you should also add support for Python3 since twodict supports both Python2 & 3.

MrS0m30n3 commented:
Regarding the **debian** directory i have no problem to keep it the main directory as long as we use a different branch.

MrS0m30n3 commented:
I also created a new branch `debian-packaging`. Now you should be able to push your changes there.

nodiscc committed:
add python3 support for debian package build, using pybuild
https://wiki.debian.org/Python/LibraryStyleGuide?action=show&redirect=Python#debian.2Frules

nodiscc commented:
Edit: fixed

~~Thanks, I made the required changes in https://github.com/MrS0m30n3/twodict/pull/1/commits/00fee07219970097a760d51881a5a5082433bc17 , but I am getting warnings `unknown substitution variable ${...:Depends}`, there is no `python3-twodict` package produced, and lintian complains about the python2 package. I'm attaching the log, at this point I have no clue but will investigate, please let me know if you have more info on how to fix this~~

nodiscc commented:
My bad, I actually had edited the wrong target in `debian/rules`. Building the package seems to work, no `python3-twodict` package is produced, I will check if this is expected. The following lintian warnings have appeared, I will try to deal with them ASAP

```
E: python-twodict: package-installs-python-pycache-dir usr/lib/python3.5/dist-packages/__pycache__/
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/__pycache__/ usr/lib/python3/dist-packages/__pycache__/
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/__pycache__/twodict.cpython-35.pyc usr/lib/python3/dist-packages/__pycache__/twodict.cpython-35.pyc
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/twodict-1.2.egg-info usr/lib/python3/dist-packages/twodict-1.2.egg-info
W: python-twodict: python-module-in-wrong-location ... use --no-tag-display-limit to see all (or pipe to a file/program)
```

MrS0m30n3 commented:
`package-installs-python-pycache-dir` the main reason for this (probably), is that twodict uses the `distutils` module instead of the newer `setuptools` in the **setup.py** file.

I guess we could add a rule in the `debian/rules` file to remove the **\_\_pycache\_\_** dir before packaging.

nodiscc commented:
https://lintian.debian.org/tags/package-installs-python-pycache-dir.html:
> The package installs a `__pycache__` directory, which is normally only used to store compiled Python source files. Compiled Python source files must not be included in the package, instead they should be generated at installation time in the postinst.

I'll try to find a similar debian package to see how this is done.

MrS0m30n3 commented:
Adding the rule below in **debian/rules** seems to work. Not sure if this is the proper way to do it though.

```make
override_dh_lintian:
    rm -rfv debian/python-twodict/usr/lib/python3.5/dist-packages/__pycache__/
    dh_lintian
```

```
$ lintian python-twodict_1.2_all.deb 
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/twodict-1.2.egg-info usr/lib/python3/dist-packages/twodict-1.2.egg-info
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/twodict.py usr/lib/python3/dist-packages/twodict.py
```

legrostdg commented:
Damn... https://salsa.debian.org/felix-guest/twodict
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=886730

MrS0m30n3 commented:
Do the tests actually pass?? When i run `dpkg-buildpackage -rfakeroot` i get:
```
...
Ran 0 test in 0.000s
...
```

legrostdg commented:
They do, in my packaging. The package is in the NEW queue, so it should be available relatively soon in Debian unstable.

MrS0m30n3 commented:
What commands you use to build the package?

legrostdg commented:
gbp buildpackage

MrS0m30n3 commented:
Again, the tests won't run for me `Ran 0 test...`.
Are you sure **all** tests run?

legrostdg commented:
You're sure you used  https://salsa.debian.org/felix-guest/twodict.git ?

legrostdg commented:
Woops, I meant https://anonscm.debian.org/git/python-modules/packages/twodict.git

legrostdg commented:
(I had to move the repo there to be hosted by the Debian Python Packaging Team, but the first repo should be uptodate, too)

legrostdg commented:
I'll just remove the repo on salsa to avoid confusion.

MrS0m30n3 commented:
I was using the repo on *salsa*, that was the problem. On the *anonscm* repo everything runs fine.

MrS0m30n3 commented:
Anyway, closing. Great job to both of you!
